### DISCUSSION TITLE
String/Memory with Max Length

### DISCUSSION THREAD

**Discussion by ackava** [dotnet/csharplang#6134] [Language Ideas]
Both string (characters array) and binary (byte array) are declared without length, however while storing the information, we do need some limits, for example column in sql table would restrict size to 50 nvarchar. 

Currently we are creating type like...

```c#
public interface IStringValue {
      string Value { get; }
}

public readonly struct String50: IStringValue {
    public readonly string Value;
    public String50(string text) {
        this.Value = text.Length <= 50 ? text : throw new ArgumentException($"Text should be less than 50 chars");
    }

    // this is so we can assign other StringValue with different sizes
    string IStringValue.Value => Value;

    // ... '+' operator will also ensure that the length is below 50
    public static operator String50 +(String50 left, String50 right) {
         return new String50(left.Value + right.Value); 
    }

    public static implicit operator String50(string text) => new String50(text);
}

// typical Entity model
[Table("customers")]
public class Customer {
     public String50 FirstName {get;set; }
}
```

So far we can create types, but for lot of variable length, it becomes a big issue.

If there was way to create such types easily with length being variable with some sort of **generic constant**. For example,

```c#
public readonly struct FixedString<N> 
    where N: const int // <-- this will force N to be a compile time constant
{
   ....
}

// typical Entity model
[Table("customers")]
public class Customer {
     public FixedString<50> FirstName {get;set; }
}
```

Internally it would not require a complicated type system, in fact, let FixedString<50> be just a syntax sugar for FixedString_50.

This will help in removing various bugs...  as we will be able to catch the exception or truncate while it is being set, before being sent to API layer for serialization.

**Comment by svick** [comment]
This is effectively a duplicate of https://github.com/dotnet/csharplang/discussions/5363.

> Internally it would not require a complicated type system, in fact, let FixedString<50> be just a syntax sugar for FixedString_50.

I think it would have to be fairly complicated. For example, consider that `FixedString<T>` is in some library and you use `FixedString<50>` in your app. How would the type `FixedString_50` be generated? And how would this work for generic methods?
